### PR TITLE
Implement Wolfram's Root builtin

### PR DESCRIPTION
**Pull request by GarkGarcia**
This is a follow up to https://github.com/mathics/Mathics/issues/801, an implementation of Wolfram's [`Root`](https://reference.wolfram.com/language/ref/Root.html). The implementation is pretty simple, it's a wrapper for SymPy's `CRootOf`.

If SymPy can represent the `CRootOf` instance with radicals then `Root` resolves to the radical representation:

```
>> Root[x ^ 2 - 1, 1]
 = -1
```

Otherwise, `Root` resolves to itself:

```
>> Root[x ^ 5 + 2 x + 1, 1]
 = Root[x ^ 5 + 2 x + 1, 1]
```

I think a got everything working, but I was not able to test my implementation. I'm not sure how testing works in this project.

## Relevant Links:

* **The official documentation of `CRootOf`:** https://docs.sympy.org/latest/modules/solvers/solvers.html#algebraic-equations
* **An explanation of `CRootOf`:** https://stackoverflow.com/a/36847486
* **Why this implementation is relevant:** https://github.com/mathics/Mathics/issues/801#issuecomment-679282075

### CODE REVIEW

**Comment by GarkGarcia** [comment]
@weakit @sn6uv @bnjones I'd like to get your input on this.

**Comment by GarkGarcia** [comment]
@sn6uv @bnjones It would be great if I could make it so that `Root[f, i]` gets converted to `CRootOf(f, i)` when the `to_sympy` method is called. I'm not sure how to do that though.

File: mathics/builtin/calculus.py [hunk 1]
```diff
@@ -534,9 +534,9 @@ class Root(Builtin):
     """
 
     messages = {
-        'nuni': "`1` is not a univariate polynomial",
-        'nint': "`1` is not an integer",
-        'iidx': "`1` is out of bounds"
+        'nuni': "Argument `1` at position 1 is not a univariate polymonial",
```
Comment (GarkGarcia) [inline mathics/builtin/calculus.py:537]: There's a typo in here

Comment (weakit) [reply mathics/builtin/calculus.py:537]: > @weakit @sn6uv @bnjones I'd like to get your input on this.

I'm not really familiar with how mathics works, and I'm not that good at working with big projects. I can help with bugs, but that's the most I think I'll be able to do.

**Comment by weakit** [comment]
> @weakit @sn6uv @bnjones I'd like to get your input on this.

I'm not really familiar with how mathics works, and I'm not that good at working with big projects. I can help with bugs, but that's the most I'll be able to do. You'll have to

**Comment by wolfv** [comment]
Hi @GarkGarcia cool!
The tests in mathics are in the documentation ;)

**Comment by GarkGarcia** [comment]
@wolfv It looks like all tests are passing, but for some reason the `Root` function isn't processed at all. I thought adding a new builtin was as simple as creating a `Builtin` subclass and implementing `apply`, is there something else I need to do? Could you help me figure this out?

File: mathics/builtin/calculus.py (hunk 1 above)
Comment (GarkGarcia) [reply mathics/builtin/calculus.py]: Fixed in https://github.com/mathics/Mathics/pull/806/commits/9c53aed4bd0de82d97d2e2f944927d4e3ab10105

File: mathics/builtin/calculus.py [hunk 2]
```diff
@@ -516,6 +516,46 @@ def apply(self, f, xs, evaluation):
         return result
 
 
+class Root(Builtin):
+    """
+    <dl>
```
Comment (wolfv) [inline mathics/builtin/calculus.py:521]: are these xml tags correct? there is maybe some missing closing tags?

Comment (wolfv) [reply mathics/builtin/calculus.py:521]: actually i just compared it, seems fine

Comment (GarkGarcia) [reply mathics/builtin/calculus.py:521]: Yeah, I was basing this in the other docs.

**Comment by GarkGarcia** [comment]
`Solve[eq, x]` is now working fine, but `Root[f, i]` isn't processed with the `Root` class for some reason. For example, `Root[x, 1]` yields `Root[x, 1]` instead of `0`. I was able to verified that `Root.apply` doesn't get called by placing a random `print` call at the beginning of `Root.apply` (the text never gets print).

**Comment by GarkGarcia** [comment]
> `Solve[eq, x]` is now working fine, but `Root[f, i]` isn't processed with the `Root` class for some reason. For example, `Root[x, 1]` yields `Root[x, 1]` instead of `0`. I was able to verified that `Root.apply` doesn't get called by placing a random `print` call at the beginning of `Root.apply` (the text never gets print).

Those are the only tests that fail.

**Comment by wolfv** [comment]
is it possible that another version of root is implemented somewhere else? Did you CTRL+F for it?

**Comment by GarkGarcia** [comment]
> Did you CTRL+F for it?

I haven't. Just tried pressing `Ctrl+F` in the command line interface and nothing happened. Is this what I'm supposed to do?

**Comment by wolfv** [comment]
ah sorry, I meant to search the entire source code for "Root" (CTRL+F on linux is the shortcut to "search-in-page").

**Comment by GarkGarcia** [comment]
> ah sorry, I meant to search the entire source code for "Root" (CTRL+F on linux is the shortcut to "search-in-page").

This is the output of searching for "Root" with `grep`

```
$ grep -E "Root" mathics/**.py
mathics/builtin/arithmetic.py:class CubeRoot(Builtin):
mathics/builtin/arithmetic.py:    <dt>'CubeRoot[$n$]'
mathics/builtin/arithmetic.py:    >> CubeRoot[16]
mathics/builtin/arithmetic.py:    #> CubeRoot[-5]
mathics/builtin/arithmetic.py:    #> CubeRoot[-510000]
mathics/builtin/arithmetic.py:    #> CubeRoot[-5.1]
mathics/builtin/arithmetic.py:    #> CubeRoot[b]
mathics/builtin/arithmetic.py:    #> CubeRoot[-0.5]
mathics/builtin/arithmetic.py:    #> CubeRoot[3 + 4 I]
mathics/builtin/arithmetic.py:        'CubeRoot[n_?NumericQ]': 'If[n > 0, Power[n, Divide[1, 3]], Times[-1, Power[Times[-1, n], Divide[1, 3]]]]',
mathics/builtin/arithmetic.py:        'CubeRoot[n_]': 'Power[n, Divide[1, 3]]',
mathics/builtin/arithmetic.py:        'CubeRoot[n_Complex]'
mathics/builtin/arithmetic.py:        evaluation.message('CubeRoot', 'preal', n)
mathics/builtin/calculus.py:     = RootSum[625 #1 ^ 4 + 125 #1 ^ 3 + 25 #1 ^ 2 + 5 #1 + 1&, Log[x + 5 #1] #1&] + Log[1 + x] / 5
mathics/builtin/calculus.py:class Root(Builtin):
mathics/builtin/calculus.py:    <dt>'Root[$f$, $i$]'
mathics/builtin/calculus.py:    >> Root[-1 + x ^ 2, 1]
mathics/builtin/calculus.py:    >> Root[-1 + x ^ 2, 2]
mathics/builtin/calculus.py:    Roots that can't be represented by radicals:
mathics/builtin/calculus.py:    >> Root[1 + 2 x + x ^ 5, 2]
mathics/builtin/calculus.py:     = Root[1 + 2 x + x ^ 5, 2]
mathics/builtin/calculus.py:        'Root[f, i]'
mathics/builtin/calculus.py:            r = sympy.CRootOf(f.to_sympy(), i.to_sympy() - 1)
mathics/builtin/calculus.py:            evaluation.message('Root', 'nuni', f)
mathics/builtin/calculus.py:            evaluation.message('Root', 'nint', i)
mathics/builtin/calculus.py:            evaluation.message('Root', 'iidx', i)
mathics/builtin/calculus.py:class FindRoot(Builtin):
mathics/builtin/calculus.py:    <dt>'FindRoot[$f$, {$x$, $x0$}]'
mathics/builtin/calculus.py:    <dt>'FindRoot[$lhs$ == $rhs$, {$x$, $x0$}]'
mathics/builtin/calculus.py:    'FindRoot' uses Newton\'s method, so the function of interest should have a first derivative.
mathics/builtin/calculus.py:    >> FindRoot[Cos[x], {x, 1}]
mathics/builtin/calculus.py:    >> FindRoot[Sin[x] + Exp[x],{x, 0}]
mathics/builtin/calculus.py:    >> FindRoot[Sin[x] + Exp[x] == Pi,{x, 0}]
mathics/builtin/calculus.py:    'FindRoot' has attribute 'HoldAll' and effectively uses 'Block' to localize $x$.
mathics/builtin/calculus.py:    >> FindRoot[Tan[x] + Sin[x] == Pi, {x, 1}]
mathics/builtin/calculus.py:    'FindRoot' stops after 100 iterations:
mathics/builtin/calculus.py:    >> FindRoot[x^2 + x + 1, {x, 1}]
mathics/builtin/calculus.py:    >> FindRoot[x ^ 2 + x + 1, {x, -I}]
mathics/builtin/calculus.py:    >> FindRoot[f[x] == 0, {x, 0}]
mathics/builtin/calculus.py:     = FindRoot[f[x] - 0, {x, 0}]
mathics/builtin/calculus.py:    >> FindRoot[Sin[x] == x, {x, 0}]
mathics/builtin/calculus.py:     = FindRoot[Sin[x] - x, {x, 0}]
mathics/builtin/calculus.py:    #> FindRoot[2.5==x,{x,0}]
mathics/builtin/calculus.py:        'FindRoot[lhs_ == rhs_, {x_, xs_}]': 'FindRoot[lhs - rhs, {x, xs}]',
mathics/builtin/calculus.py:        'FindRoot[f_, {x_, x0_}]'
mathics/builtin/calculus.py:            evaluation.message('FindRoot', 'snum', x0)
mathics/builtin/calculus.py:            evaluation.message('FindRoot', 'sym', x, 2)
mathics/builtin/calculus.py:                evaluation.message('FindRoot', 'dsing', x, x0)
mathics/builtin/calculus.py:                evaluation.message('FindRoot', 'nnum', x, x0)
mathics/builtin/calculus.py:            evaluation.message('FindRoot', 'maxiter')
mathics/builtin/files.py:class RootDirectory(Predefined):
mathics/builtin/files.py:    <dt>'$RootDirectory'
mathics/builtin/files.py:    >> $RootDirectory
mathics/builtin/files.py:    name = '$RootDirectory'
mathics/builtin/files.py:    #> FileNameDepth[$RootDirectory]
mathics/core/convert.py:    elif isinstance(expr, sympy.RootSum):
mathics/core/convert.py:        return Expression('RootSum', from_sympy(expr.poly),
mathics/core/convert.py:    elif isinstance(expr, sympy.CRootOf):
mathics/core/convert.py:        return Expression('Root', from_sympy(e), i + 1)
```

In other words, `Root` isn't defined anywhere else.

**Comment by GarkGarcia** [comment]
@wolfv Could you take a look at the code? I don't understand why `Root[f, i]` doesn't work.

**Comment by GarkGarcia** [comment]
Just found out what the error was. As one might expect, it's one of those silly errors that are a pain to debug.

Apparently the way Mathics patterns are generated from builtin classes is via the docstring of their `apply` method implementation. The docstring for `Root.apply` was `'Root[f, i]'` instead of `'Root[f_, i_]'`.

**Comment by GarkGarcia** [comment]
Apparently there's only one test that isn't passing (which has nothing to do with this PR). The only thing left to address is that fact that Wolfram's `Root` uses that _slot_ thing (and I don't know what it is) and our `Root` uses regular unbound variables.

* **Wolfram:** `Root[#1 ^ 5 + 2 #1 + 1 &, 2]`
* **This implementation:** `Root[x ^ 5 + 2 x + 1, 2]`

I don't know how much of _slot_ is implemented in Mathics and how could I convert a regular (univariate) equation to this format.

**Comment by GarkGarcia** [comment]
@weakit Perhaps you could help me understand what _slot_ is all about.

**Comment by suhr** [comment]
https://reference.wolfram.com/language/ref/Function.html

**Comment by weakit** [comment]
> 
> 
> @weakit Perhaps you could help me understand what _slot_ is all about.

K I'll try my best.

**Comment by weakit** [comment]
> 
> 
> @weakit Perhaps you could help me understand what _slot_ is all about.

Slot is used to represent the arguments of a function.

So say you have a function f(x, y) = x² + y + 1

In Mathematica, you could represent the function as `(#1^2 + #2 + 1)&`, where `#1` and `#2` are the first and second arguments.

if `f` is `#1^2 + #2 + 1`
`f[x, y]` would work out to `x^2 + y + 2` 

Root represents the function `x^5 + 2 x + 1` as `#^5 + 2 # + 1`.

I remember that mathics did have Slot[], but I don't remember if it worked properly. Try looking for it.

This is the way I understand it. It may be a bit off (?) so make sure to go through the documentation once.

Try going through these:
https://reference.wolfram.com/language/howto/WorkWithPureFunctions.html
https://reference.wolfram.com/language/ref/Function.html

**Comment by GarkGarcia** [comment]
> > @weakit Perhaps you could help me understand what _slot_ is all about.
> 
> Slot is used to represent the arguments of a function.
> 
> So say you have a function f(x, y) = x² + y + 1,
> In Mathematica, you could represent the function as `(#1^2 + #2 + 1)&`, where `#1` and `#2` are the first and second arguments.
> 
> if `f` is `#1^2 + #2 + 1`
> `f[x, y]` would work out to `x^2 + y + 2`
> 
> Root represents the function `x^5 + 2 x + 1` as `#^5 + 2 # + 1`.
> 
> I remember that mathics did have Slot[], but I don't remember if it worked properly. Try looking for it.
> 
> This is the way I understand it. It may be a bit off (?) so make sure to go through the documentation once.
> 
> Try going through these:
> https://reference.wolfram.com/language/howto/WorkWithPureFunctions.html
> https://reference.wolfram.com/language/ref/Function.html

I see. @wolfv How do I appropriately convert between a `Function` and a `sympy.PurePolly`?

**Comment by suhr** [comment]
The most straightforward way it to encode `#1` as a variable with name `_1`, and so on.

**Comment by GarkGarcia** [comment]
> The most straightforward way it to encode `#1` as a variable with name `_1`, and so on.

Seems a bit hacky but I works well enough 😋️

Alright, I've made it so that `Root` takes a proper function (with slots) as it's first argument. If the implementation passes the tests we're all done in here.

**Comment by suhr** [comment]
You need to fix the tests though.

File: mathics/builtin/calculus.py [hunk 3]
```diff
@@ -524,14 +524,14 @@ class Root(SympyFunction):
...
      = -1
-    >> Root[-1 + x ^ 2, 2]
+    >> Root[-1 + #1 ^ 2&, 2]
      = 1
 
     Roots that can't be represented by radicals:
-    >> Root[1 + 2 x + x ^ 5, 2]
-     = Root[1 + 2 x + x ^ 5, 2]
+    >> Root[1 + 2 #1 + #1 ^ 5&, 2]
+     = Root[1 + 2 #1 + #1 ^ 5&, 2]
```
Comment (suhr) [inline mathics/builtin/calculus.py]:
```suggestion
    >> Root[(1 + 2 #1 + #1 ^ 5) &, 2]
     = Root[(1 + 2 #1 + #1 ^ 5) &, 2]
```

**Comment by GarkGarcia** [comment]
@suhr The `(#1 ...) &` idea didn't work out, it caused the tests to fail because the expected results where not format in the way Mathics prints them (although both expressions represent the same value).

**Comment by GarkGarcia** [comment]
It looks like all relevant tests are passing. @wolfv Can we merge the PR?

**Comment by wolfv** [comment]
great! thanks!!

**Pull request merged**